File: lib/screens/survey_screen.dart
// lib/screens/survey_screen.dart
import 'package:flutter/material.dart';
import '../config/theme.dart';
import 'package:provider/provider.dart';
import '../models/asset_model.dart';
import '../providers/auth_provider.dart';
import '../providers/asset_provider.dart';
import '../providers/theme_provider.dart';
import '../services/rbac_service.dart';
import '../configs/routes.dart';
import '../widgets/cost_center_selector.dart';
import '../widgets/asset_class_picker.dart';
import '../widgets/asset_search_bar.dart';
import '../widgets/load_more_list.dart';
import '../widgets/image_modal.dart';
import 'audit_screen.dart';
import 'temp_photo_screen.dart';

class SurveyScreen extends StatefulWidget {
  const SurveyScreen({super.key});

  @override
  State<SurveyScreen> createState() => _SurveyScreenState();
}

class _SurveyScreenState extends State<SurveyScreen> {
  AssetModel? _selectedAsset;
  String _searchQuery = '';
  String? _selectedCostCenter;
  String? _selectedAssetClass;
  bool _showUnauditedOnly = true;

  @override
  Widget build(BuildContext context) {
    final auth = context.watch<AuthProvider>();
    final assetProv = context.watch<AssetProvider>();
    final themeProvider = context.watch<ThemeProvider>(); // 🟢 เรียกใช้เฝ้าดูสถานะธีม

    if (auth.isAppLoading) {
      return const Scaffold(
        body: Center(child: CircularProgressIndicator()),
      );
    }

    if (auth.user == null) {
      return Scaffold(
        appBar: AppBar(title: const Text('Asset Survey')),
        body: Center(
          child: Card(
            margin: const EdgeInsets.all(24),
            child: Padding(
              padding: const EdgeInsets.all(24),
              child: Column(
                mainAxisSize: MainAxisSize.min,
                children: [
                  const Text('กรุณาเข้าสู่ระบบ',
                      style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold)),
                  const SizedBox(height: 8),
                  const Text('ระบบจำเป็นต้องยืนยันตัวตนผ่านบัญชี Google'),
                  const SizedBox(height: 16),
                  ElevatedButton.icon(
                    onPressed: () => auth.login(),
                    icon: const Icon(Icons.login),
                    label: const Text('Sign in with Google'),
                  ),
                ],
              ),
            ),
          ),
        ),
      );
    }

    if (!auth.authorized) {
      return Scaffold(
        appBar: AppBar(title: const Text('Asset Survey')),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              const Text('ไม่มีสิทธิ์เข้าใช้งาน',
                  style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold)),
              Text('บัญชี ${auth.user!.email} ไม่อยู่ในรายชื่อผู้มีสิทธิ์'),
              const SizedBox(height: 16),
              ElevatedButton(
                onPressed: () => auth.logout(),
                child: const Text('ออกจากระบบ'),
              ),
            ],
          ),
        ),
      );
    }

    final allowedCostCenters = AppRoutes.getAllowedCostCentersV2(
      screenName: AppRoutes.survey,
      role: auth.role,
      userCostCenters: auth.allowedCostCenters,
    );

    final availableCostCenters = RbacService.getAvailableCostCenters(
      assetProv.assets,
      RBACContext(
        role: auth.role,
        allowedCostCenters: allowedCostCenters,
      ),
    );

    if (_selectedCostCenter == null && allowedCostCenters is List && allowedCostCenters!.isNotEmpty) {
      WidgetsBinding.instance.addPostFrameCallback((_) {
        setState(() => _selectedCostCenter = allowedCostCenters[0]);
      });
    }

    final visibleAssets = RbacService.filterAssets(
      assetProv.assets,
      RBACContext(
        role: auth.role,
        allowedCostCenters: allowedCostCenters,
      ),
    );

    var filteredAssets = visibleAssets;
    if (_selectedCostCenter != null) {
      filteredAssets = filteredAssets.where((a) => a.costCenter == _selectedCostCenter).toList();
    }
    if (_selectedAssetClass != null) {
      filteredAssets = filteredAssets.where((a) => a.assetClass == _selectedAssetClass).toList();
    }
    if (_showUnauditedOnly) {
      filteredAssets = filteredAssets.where((a) => !assetProv.auditedAssetNos.contains(a.assetNo)).toList();
    }
    if (_searchQuery.isNotEmpty) {
      final q = _searchQuery.toUpperCase();
      filteredAssets = filteredAssets.where((a) =>
        a.assetNo.toUpperCase().contains(q) ||
        a.description.toUpperCase().contains(q) ||
        a.lastLocationName.toUpperCase().contains(q) ||
        a.mainLocation.toUpperCase().contains(q) ||
        a.costCenter.toUpperCase().contains(q) ||
        a.costCenterName.toUpperCase().contains(q) ||
        a.assetOwner.toUpperCase().contains(q) ||
        (a.remarks?.toUpperCase() ?? '').contains(q)
      ).toList();
    }

    final poolForClasses = _selectedCostCenter != null
        ? visibleAssets.where((a) => a.costCenter == _selectedCostCenter).toList()
        : visibleAssets;
    final classMap = <String, AssetClassStats>{};
    for (final a in poolForClasses) {
      classMap.putIfAbsent(a.assetClass, () => AssetClassStats(
        assetClass: a.assetClass,
        assetClassName: a.assetClassName,
        total: 0,
        audited: 0,
      )).total++;
    }
    final availableAssetClasses = classMap.values.toList()
      ..sort((a, b) => a.assetClass.compareTo(b.assetClass));

    // ✅ ตรวจสอบว่า _selectedAssetClass ยังมีอยู่ใน available classes หรือไม่
    if (_selectedAssetClass != null && 
        !availableAssetClasses.any((ac) => ac.assetClass == _selectedAssetClass)) {
      _selectedAssetClass = null;
    }

    final isTablet = MediaQuery.of(context).size.shortestSide >= 600;

    return Scaffold(
      appBar: AppBar(
        title: const Text('ASSET DEV — Survey'),
        actions: isTablet
            // ── iPad: แสดงปุ่มทั้งหมด ──
            ? [
                IconButton(
                  tooltip: 'Search',
                  icon: const Icon(Icons.search),
                  onPressed: () => Navigator.pushNamed(context, AppRoutes.search),
                ),
                IconButton(
                  tooltip: 'Dashboard',
                  icon: const Icon(Icons.dashboard),
                  onPressed: () => Navigator.pushNamed(context, AppRoutes.dashboard),
                ),
                IconButton(
                  tooltip: 'Temp Photos',
                  icon: const Icon(Icons.camera_alt_outlined),
                  onPressed: () => _showTempPhotosPage(context),
                ),
                IconButton(
                  tooltip: themeProvider.isDarkMode ? 'Light mode' : 'Dark mode',
                  icon: Icon(themeProvider.isDarkMode ? Icons.light_mode : Icons.dark_mode),
                  onPressed: () {
                    final t = context.read<ThemeProvider>();
                    t.setThemeMode(t.isDarkMode ? ThemeMode.light : ThemeMode.dark);
                  },
                ),
                // Avatar
                CircleAvatar(
                  radius: 16,
                  backgroundImage: auth.user?.photoURL != null
                      ? NetworkImage(auth.user!.photoURL!)
                      : null,
                  backgroundColor: context.primary,
                  child: auth.user?.photoURL == null
                      ? Text(
                          (auth.user?.email ?? 'U')[0].toUpperCase(),
                          style: const TextStyle(color: Colors.white, fontSize: 14, fontWeight: FontWeight.bold),
                        )
                      : null,
                ),
                const SizedBox(width: 4),
              ]
            // ── iPhone: PopupMenuButton (เหมือนเดิม) ──
            : [
                PopupMenuButton<String>(
                  onSelected: (route) {
                    if (route == AppRoutes.dashboard) {
                      Navigator.pushNamed(context, AppRoutes.dashboard);
                    } else if (route == AppRoutes.search) {
                      Navigator.pushNamed(context, AppRoutes.search);
                    } else if (route == 'temp_photos') {
                      _showTempPhotosPage(context);
                    } else if (route == 'toggle_theme') {
                      final theme = context.read<ThemeProvider>();
                      theme.setThemeMode(theme.isDarkMode ? ThemeMode.light : ThemeMode.dark);
                    } else if (route == 'logout') {
                      context.read<AuthProvider>().logout();
                    }
                  },
                  child: Padding(
                    padding: const EdgeInsets.symmetric(horizontal: 8.0),
                    child: CircleAvatar(
                      radius: 16,
                      backgroundImage: auth.user?.photoURL != null
                          ? NetworkImage(auth.user!.photoURL!)
                          : null,
                      backgroundColor: context.primary,
                      child: auth.user?.photoURL == null
                          ? Text(
                              (auth.user?.email ?? 'U')[0].toUpperCase(),
                              style: const TextStyle(color: Colors.white, fontSize: 14, fontWeight: FontWeight.bold),
                            )
                          : null,
                    ),
                  ),
                  itemBuilder: (_) => [
                    PopupMenuItem<String>(
                      enabled: false,
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Text(auth.user?.displayName ?? 'User', style: const TextStyle(fontWeight: FontWeight.bold, fontSize: 14)),
                          Text(auth.user?.email ?? '', style: TextStyle(fontSize: 11, color: context.textSecondary)),
                        ],
                      ),
                    ),
                    const PopupMenuDivider(),
                    const PopupMenuItem(value: AppRoutes.dashboard, child: Text('📊 Dashboard')),
                    const PopupMenuItem(value: AppRoutes.search, child: Text('🔍 Search')),
                    const PopupMenuItem(value: 'temp_photos', child: Text('📸 Temp Photos')),
                    const PopupMenuDivider(),
                    PopupMenuItem(
                      value: 'toggle_theme',
                      child: Row(
                        children: [
                          Icon(themeProvider.isDarkMode ? Icons.light_mode : Icons.dark_mode, size: 18),
                          const SizedBox(width: 8),
                          Text(themeProvider.isDarkMode ? '☀️ Light mode' : '🌙 Dark mode'),
                        ],
                      ),
                    ),
                    const PopupMenuDivider(),
                    const PopupMenuItem(value: 'logout', child: Text('🚪 Sign out', style: TextStyle(color: Colors.red))),
                  ],
                ),
        ],
      ),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(12),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            LinearProgressIndicator(
              value: assetProv.totalCount > 0
                  ? assetProv.auditedCount / assetProv.totalCount
                  : 0,
            ),
            const SizedBox(height: 8),
            Text(
              '${assetProv.auditedCount}/${assetProv.totalCount} สำรวจแล้ว',
              style: const TextStyle(fontSize: 12),
            ),
            const SizedBox(height: 12),

            CostCenterSelector(
              costCenters: availableCostCenters,
              selectedCostCenter: _selectedCostCenter,
              onSelect: (cc) => setState(() {
                _selectedCostCenter = cc;
                _selectedAssetClass = null;
              }),
              auditedCounts: _buildAuditedCounts(visibleAssets, assetProv.auditedAssetNos),
              hideAll: allowedCostCenters is List,
            ),
            const SizedBox(height: 12),

            if (_selectedCostCenter != null)
              AssetClassPicker(
                classes: availableAssetClasses,
                selectedClass: _selectedAssetClass,
                onSelect: (ac) => setState(() => _selectedAssetClass = ac),
                auditedCounts: _buildClassAuditedCounts(poolForClasses, assetProv.auditedAssetNos),
              ),
            if (_selectedCostCenter != null) const SizedBox(height: 12),

            Row(
              children: [
                Switch(
                  value: _showUnauditedOnly,
                  onChanged: (v) => setState(() => _showUnauditedOnly = v),
                ),
                Text(_showUnauditedOnly ? 'เหลือยังไม่ตรวจ' : 'ทั้งหมด'),
                const Spacer(),
                Text('${filteredAssets.length} รายการ', style: const TextStyle(fontSize: 11)),
              ],
            ),

            AssetSearchBar(
              value: _searchQuery,
              onChanged: (v) => setState(() => _searchQuery = v),
            ),

            LoadMoreList(
              assets: filteredAssets,
              selectedAssetNo: _selectedAsset?.assetNo,
              onSelect: (asset) => Navigator.push(
                context,
                MaterialPageRoute(
                  builder: (_) => AuditScreen(asset: asset),
                ),
              ),
              onImageClick: (url) => showImageModal(context, url),
              auditedSet: assetProv.auditedAssetNos,
              pageSize: 50,
            ),

            const SizedBox(height: 32),
          ],
        ),
      ),
    );
  }

  void _showTempPhotosPage(BuildContext context) {
    Navigator.push(
      context,
      MaterialPageRoute(
        builder: (_) => const TempPhotoScreen(),
      ),
    );
  }

  Map<String, int> _buildAuditedCounts(List<AssetModel> assets, Set<String> auditedSet) {
    final map = <String, int>{};
    for (final a in assets) {
      if (auditedSet.contains(a.assetNo)) {
        map[a.costCenter] = (map[a.costCenter] ?? 0) + 1;
      }
    }
    return map;
  }

  Map<String, int> _buildClassAuditedCounts(List<AssetModel> assets, Set<String> auditedSet) {
    final map = <String, int>{};
    for (final a in assets) {
      if (auditedSet.contains(a.assetNo)) {
        map[a.assetClass] = (map[a.assetClass] ?? 0) + 1;
      }
    }
    return map;
  }
}

File: AssetApp-Android/app/src/main/java/com/plbg/assetapp/ui/survey/SurveyScreen.kt
package com.plbg.assetapp.ui.survey

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import com.plbg.assetapp.domain.model.AppRoutes
import com.plbg.assetapp.domain.model.AssetModel
import com.plbg.assetapp.domain.model.RBACContext
import com.plbg.assetapp.domain.usecase.RbacService
import com.plbg.assetapp.ui.auth.AuthUser
import com.plbg.assetapp.ui.auth.AuthViewModel
import com.plbg.assetapp.ui.common.AssetSearchBar
import com.plbg.assetapp.ui.common.LoadMoreList

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SurveyScreen(
    viewModel: AssetViewModel = hiltViewModel(),
    authViewModel: AuthViewModel = hiltViewModel(),
    onNavigateToAudit: (AssetModel) -> Unit,
    onNavigateToSearch: () -> Unit,
    onNavigateToDashboard: () -> Unit,
    onNavigateToTempPhotos: () -> Unit,
) {
    val uiState by viewModel.uiState.collectAsState()
    val authState by authViewModel.uiState.collectAsState()
    val auditedAssetNos by viewModel.auditedAssetNos.collectAsState()

    var selectedCostCenter by rememberSaveable { mutableStateOf<String?>(null) }
    var selectedAssetClass by rememberSaveable { mutableStateOf<String?>(null) }
    var searchQuery by rememberSaveable { mutableStateOf("") }
    var showUnauditedOnly by rememberSaveable { mutableStateOf(true) }

    when (val state = authState) {
        is AuthViewModel.UiState.Loading -> {
            Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                CircularProgressIndicator()
            }
            return
        }
        is AuthViewModel.UiState.NotLoggedIn -> {
            com.plbg.assetapp.ui.auth.LoginScreen(
                onLogin = { authViewModel.login() }
            )
            return
        }
        is AuthViewModel.UiState.Unauthorized -> {
            com.plbg.assetapp.ui.auth.UnauthorizedScreen(
                email = state.email,
                onLogout = { authViewModel.logout() }
            )
            return
        }
        is AuthViewModel.UiState.LoggedIn -> { }
    }

    val authUser = (authState as AuthViewModel.UiState.LoggedIn).user
    val assets = when (val state = uiState) {
        is AssetViewModel.UiState.Success -> state.assets
        else -> emptyList()
    }

    val allowedCostCenters = AppRoutes.getAllowedCostCenters(
        screenName = AppRoutes.SURVEY,
        role = authUser.role,
        userCostCenters = authUser.allowedCostCenters,
    )

    val visibleAssets = remember(assets, allowedCostCenters, authUser.role) {
        RbacService.filterAssets(
            assets,
            RBACContext(role = authUser.role, allowedCostCenters = allowedCostCenters),
        )
    }

    LaunchedEffect(allowedCostCenters) {
        if (selectedCostCenter == null &&
            allowedCostCenters is List &&
            allowedCostCenters.isNotEmpty()) {
            selectedCostCenter = allowedCostCenters[0]
        }
    }

    val filteredAssets = remember(
        visibleAssets, selectedCostCenter, selectedAssetClass,
        searchQuery, showUnauditedOnly, auditedAssetNos
    ) {
        visibleAssets
            .filter { selectedCostCenter == null || it.costCenter == selectedCostCenter }
            .filter { selectedAssetClass == null || it.assetClass == selectedAssetClass }
            .filter { !showUnauditedOnly || !auditedAssetNos.contains(it.assetNo) }
            .filter {
                searchQuery.isEmpty() || run {
                    val q = searchQuery.uppercase()
                    it.assetNo.uppercase().contains(q) ||
                    it.description.uppercase().contains(q) ||
                    it.lastLocationName.uppercase().contains(q) ||
                    it.costCenter.uppercase().contains(q) ||
                    (it.remarks?.uppercase() ?: "").contains(q)
                }
            }
    }

    val availableCostCenters = remember(visibleAssets, authUser.role, allowedCostCenters) {
        RbacService.getAvailableCostCenters(
            visibleAssets,
            RBACContext(role = authUser.role, allowedCostCenters = allowedCostCenters),
        )
    }

    val availableAssetClasses = remember(visibleAssets, selectedCostCenter) {
        val pool = if (selectedCostCenter != null)
            visibleAssets.filter { it.costCenter == selectedCostCenter }
        else visibleAssets
        pool.groupBy { it.assetClass }.map { (ac, list) ->
            com.plbg.assetapp.domain.model.AssetClassStats(
                assetClass = ac,
                assetClassName = list.first().assetClassName,
                total = list.size,
                audited = list.count { auditedAssetNos.contains(it.assetNo) },
            )
        }.sortedBy { it.assetClass }
    }

    val isTablet = LocalConfiguration.current.smallestScreenWidthDp >= 600

    Scaffold(
        topBar = {
            SurveyTopBar(
                user = authUser,
                isTablet = isTablet,
                onSearch = onNavigateToSearch,
                onDashboard = onNavigateToDashboard,
                onTempPhotos = onNavigateToTempPhotos,
                onLogout = { authViewModel.logout() },
            )
        }
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .verticalScroll(rememberScrollState())
                .padding(12.dp),
        ) {
            val totalCount by viewModel.totalCount.collectAsState()
            val auditedCount by viewModel.auditedCount.collectAsState()

            LinearProgressIndicator(
                progress = { if (totalCount > 0) auditedCount.toFloat() / totalCount else 0f },
                modifier = Modifier.fillMaxWidth(),
            )
            Spacer(Modifier.height(8.dp))
            Text("$auditedCount/$totalCount audited", fontSize = 12.sp)

            Spacer(Modifier.height(12.dp))

            CostCenterSelector(
                costCenters = availableCostCenters,
                selectedCostCenter = selectedCostCenter,
                onSelect = { selectedCostCenter = it; selectedAssetClass = null },
                hideAll = allowedCostCenters is List,
            )

            Spacer(Modifier.height(12.dp))

            if (selectedCostCenter != null && availableAssetClasses.isNotEmpty()) {
                AssetClassPicker(
                    classes = availableAssetClasses,
                    selectedClass = selectedAssetClass,
                    onSelect = { selectedAssetClass = it },
                )
                Spacer(Modifier.height(12.dp))
            }

            Row(verticalAlignment = Alignment.CenterVertically) {
                Switch(
                    checked = showUnauditedOnly,
                    onCheckedChange = { showUnauditedOnly = it },
                )
                Text(
                    if (showUnauditedOnly) "Pending audit" else "All",
                    fontSize = 12.sp,
                )
                Spacer(Modifier.weight(1f))
                Text("${filteredAssets.size} items", fontSize = 11.sp)
            }

            AssetSearchBar(
                value = searchQuery,
                onValueChange = { searchQuery = it },
            )

            when (val state = uiState) {
                is AssetViewModel.UiState.Loading -> {
                    Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                        CircularProgressIndicator()
                    }
                }
                is AssetViewModel.UiState.Error -> {
                    Text(
                        state.message,
                        color = MaterialTheme.colorScheme.error,
                        fontSize = 14.sp,
                    )
                }
                is AssetViewModel.UiState.Success -> {
                    LoadMoreList(
                        assets = filteredAssets,
                        auditedSet = auditedAssetNos,
                        onAssetClick = onNavigateToAudit,
                        pageSize = 50,
                    )
                }
            }
        }
    }
}